src/Native/Runtime/gcrhenv.cpp

```diff
@@ -1504,6 +1416,159 @@ bool GCToEEInterface::EagerFinalized(Object* obj)
     return false;
 }
 
+bool GCToEEInterface::IsGCThread()
+{
+    return false;
+}
+
+bool GCToEEInterface::WasCurrentThreadCreatedByGC()
+{
+    return false;
+}
+
+struct SuspendableThreadStubArguments
+{
+    void (*m_pRealStartRoutine)(void*);
+    void* m_pRealContext;
+    Thread* m_pThread;
+    CLREventStatic m_ThreadStartedEvent;
+};
+
+struct ThreadStubArguments
+{
+    void* Argument;
+    void (*ThreadStart)(void*);
+    HANDLE Thread;
+    bool HasStarted;
+    CLREventStatic ThreadStartedEvent;
+};
+
+namespace
+{
+    const size_t MaxThreadNameSize = 255;
+
+    bool CreateSuspendableThread(
+        void (*threadStart)(void*),
+        void* argument,
+        const char* name)
+    {
+        SuspendableThreadStubArguments threadStubArgs;
+
+        threadStubArgs.m_pThread = NULL;
+        threadStubArgs.m_pRealStartRoutine = threadStart;
+        threadStubArgs.m_pRealContext = argument;
+
+        if (!threadStubArgs.m_ThreadStartedEvent.CreateAutoEventNoThrow(false))
+        {
+            return NULL;
+        }
+
+        // Helper used to wrap the start routine of background GC threads so we can do things like initialize the
+        // Redhawk thread state which requires running in the new thread's context.
+        auto threadStub = [](void* argument) -> DWORD
+        {
+            SuspendableThreadStubArguments* pStartContext = (SuspendableThreadStubArguments*)argument;
+
+            // Initialize the Thread for this thread. The false being passed indicates that the thread store lock
+            // should not be acquired as part of this operation. This is necessary because this thread is created in
+            // the context of a garbage collection and the lock is already held by the GC.
+            ASSERT(GCHeapUtilities::IsGCInProgress());
+            ThreadStore::AttachCurrentThread(false);
+
+            Thread* pThread = ThreadStore::GetCurrentThread();
+            pThread->SetGCSpecial(true);
+
+            // Inform the GC which Thread* we are.
+            pStartContext->m_pThread = pThread;
+
+            auto realStartRoutine = pStartContext->m_pRealStartRoutine;
+            void* realContext = pStartContext->m_pRealContext;
+
+            pStartContext->m_ThreadStartedEvent.Set();
+
+            STRESS_LOG_RESERVE_MEM(GC_STRESSLOG_MULTIPLY);
+
+            realStartRoutine(realContext);
+
+            return 0;
+        };
+
+        if (!PalStartBackgroundGCThread(threadStub, &threadStubArgs))
+        {
+            threadStubArgs.m_ThreadStartedEvent.CloseEvent();
+            return NULL;
+        }
+
+        uint32_t res = threadStubArgs.m_ThreadStartedEvent.Wait(INFINITE, FALSE);
+        threadStubArgs.m_ThreadStartedEvent.CloseEvent();
+        ASSERT(res == WAIT_OBJECT_0);
+
+        ASSERT(threadStubArgs.m_pThread != NULL);
+        return threadStubArgs.m_pThread != NULL;
+    }
+
+    bool CreateNonSuspendableThread(
```

Comment:
The only difference between suspendable and non-suspendable threads should be whether `TreadStore::AttachCurrentThread` is called for them. It should not be necessary to duplicate the whole method.

Reply:
And only suspendable threads `pThread->SetGCSpecial(true);`? (looking at [coreclr's version](https://github.com/dotnet/coreclr/blob/e64cbb5c722ab9b7d93ca32445e17df1115be70c/src/vm/gcenv.ee.cpp#L1144-L1295) which did duplicate them).

Reply:
CoreCLR calls SetGCSpecial for both suspendable and non-suspendable threads.

Reply:
Where? `ThreadStore::GetCurrentThread()` asserts `pCurThread->IsInitialized()` when `ThreadStore::AttachCurrentThread(false)` isn't called which is the case for non suspendable threads.

Reply:
The IsGCSpecial checks should use `ThreadStore::RawGetCurrentThread` to avoid the assert in regular GetCurrentThread.